clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -4071,6 +4081,168 @@ void handleFailure(Throwable throwable) {
+        runnable.call(call, now);
+        return new DescribeUserScramCredentialsResult(future);
+    }
+
+    @Override
+    public AlterUserScramCredentialsResult alterUserScramCredentials(List<UserScramCredentialAlteration> alterations,
+                                                                     AlterUserScramCredentialsOptions options) {
+        final long now = time.milliseconds();
+        final Map<String, KafkaFutureImpl<Void>> futures = new HashMap<>();
+        for (UserScramCredentialAlteration alteration: alterations) {
+            futures.put(alteration.getUser(), new KafkaFutureImpl<>());
+        }
+        final Map<String, Exception> userIllegalAlterationExceptions = new HashMap<>();
+        // We need to keep track of users with deletions of an unknown SCRAM mechanism
+        alterations.stream().filter(a -> a instanceof UserScramCredentialDeletion).forEach(alteration -> {
+            UserScramCredentialDeletion deletion = (UserScramCredentialDeletion) alteration;
+            ScramMechanism mechanism = deletion.getMechanism();
+            if (mechanism == null || mechanism == ScramMechanism.UNKNOWN) {
+                userIllegalAlterationExceptions.put(deletion.getUser(), new IllegalArgumentException("Unknown SCRAM mechanism"));
+            }
+        });
+        // Creating an upsertion may throw InvalidKeyException or NoSuchAlgorithmException,
+        // so keep track of which users are affected by such a failure and immediately fail all their alterations
+        final Map<String, Map<ScramMechanism, AlterUserScramCredentialsRequestData.ScramCredentialUpsertion>> userInsertions = new HashMap<>();
+        alterations.stream().filter(a -> a instanceof UserScramCredentialUpsertion)
+                .filter(alteration -> !userIllegalAlterationExceptions.containsKey(alteration.getUser()))
+                .forEach(alteration -> {
+                    UserScramCredentialUpsertion upsertion = (UserScramCredentialUpsertion) alteration;
+                    String user = upsertion.getUser();
+                    try {
+                        ScramMechanism mechanism = upsertion.getInfo().getMechanism();
+                        if (mechanism == null || mechanism == ScramMechanism.UNKNOWN)
+                            throw new IllegalArgumentException("Unknown SCRAM mechanism");
+                        userInsertions.putIfAbsent(user, new HashMap<>());
+                        userInsertions.get(user).put(mechanism, getScramCredentialUpsertion(upsertion));
+                    } catch (Exception e) {
+                        // we might overwrite aa exception from a previous upsertion, but we don't really care
+                        // since we just needs to mark this user as having at least one illegal alteration
+                        // and make an exception instance available for completing the corresponding future exceptionally
+                        userIllegalAlterationExceptions.put(user, e);
```

Comment:
We should have a defined ApiException subclass for this.  NoSuchAlgorithmException could probably be shoehorned into INVALID_CONFIG.  InvalidKeyException probably needs its own defined error code (which we should add to the KIP).

This may seem bureaucratic, but think about it this way: maybe we'll want to do this on the broker side eventually.  Then we need a way to serialize, which plain old java exceptions don't have.

Reply:
I wonder if `InvalidRequestException` is indeed the right one.  On the broker side, in `kafka.server.AdminManager`, the code looks like this:

```
    invalidUsers.foreach(user => retval.results.add(new AlterUserScramCredentialsResult().setUser(user)
      .setErrorCode(Errors.INVALID_REQUEST.code).setErrorMessage("Unknown SCRAM mechanism or too few iterations")))
```

Reply:
The pattern in Kafka has always been to have very specific error codes and exceptions.  Unfortunately INVALID_REQUEST is documented like this:
```
"This most likely occurs because of a request being malformed by the " +
"client library or the message was sent to an incompatible broker. See the broker logs " +
"for more details.",
```
So we probably need a new error code here

Reply:
`NoSuchAlgorithmException` is thrown if we don't support the (otherwise valid) SCRAM mechanism.  For example, this might occur if we add support for a new SCRAM mechanism but the client is running an older JRE that doesn't support the algorithm associated with that newer mechanism.  `UnsupportedSaslMechanismException` would be appropriate here, I think.

`InvalidKeyException` is thrown if the password to be salted is unacceptable -- i.e. if it is empty.  I will add a new UnacceptablePasswordException and Errors.UNACCEPTABLE_PASSWORD as a starting point for further review/discussion.

Reply:
Actually, we also have to deal with too few/too many iterations as well as empty username.  These, along with empty password, are a class unto themselves -- unacceptable credentials.  So I'm abandoning the specific "`UnacceptablePasswordException`/`Errors.UNACCEPTABLE_PASSWORD`" in favor of "`UnacceptableCredentialException`/`Errors.UNACCEPTABLE_CREDENTIAL`"